test/core/gprpp/match_test.cc

```diff
@@ -0,0 +1,45 @@
+// Copyright 2021 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "src/core/lib/gprpp/match.h"
+#include <gtest/gtest.h>
+
+namespace grpc_core {
+namespace testing {
+
+TEST(MatchTest, Test) {
```

Comment:
If I'm reading the code correctly, it looks like it should allow the functions to modify the value in the variant by taking their parameters as a reference.  Is that right?  If so, how about adding a test that covers that case?

Reply:
I'd left that open and it turned out to be a terrible idea... so I removed it.

If we end up wanting it, let's re-add it later in a principled fashion (maybe `MutatingMatch(variant<>*, cases...)`

Reply:
I think we actually do want it.  I just realized that we have exactly such a case in the code today, although we were doing a copy where we should have done a move.  I've just pushed https://github.com/grpc/grpc/pull/26481/commits/769896d1018fed8bc294a0baeb8b80a7058becc8 to #26481 to fix it.

Why do you say that it turned out to be a terrible idea?  What kind of problems does this cause?

Reply:
The mutable version looks good.  Can you please add a test for that?  Thanks!